update-data job is failing

```
Debug: Populating scores for 2024-05-01
Successfully populated 0 scores
Error: Incorrect content encoding: undefined
    at IncomingMessage.<anonymous> (/home/runner/work/langtrends-data/langtrends-data/src/StackOverflow.ts:113:17)
    at IncomingMessage.emit (node:events:530:35)
    at IncomingMessage.emit (node:domain:488:12)
    at endReadableNT (node:internal/streams/readable:1696:12)
    at processTicksAndRejections (node:internal/process/task_queues:82:21)
Error: Process completed with exit code 1.
```

https://github.com/bmaupin/langtrends-data/actions/runs/8906339829/job/24474638480

--- a/scripts/calculate-deviation.ts
+++ b/scripts/calculate-deviation.ts
@@ -4,6 +4,7 @@
 
 import 'dotenv/config';
 
+import { oldestDate } from '../src/consts';
 import GitHub from '../src/GitHub';
 import settings from '../src/settings.json';
 import StackOverflow from '../src/StackOverflow';
@@ -11,8 +12,6 @@ import StackOverflow from '../src/StackOverflow';
 import languages from '../data/languages.json';
 import scores from '../data/scores-full.json';
 import { Language } from '../src/DataPopulator';
-
-const oldestDate = new Date('2008-02-01');
 
 const main = async () => {
   let currentDate = '2008-02-01';

--- a/scripts/get-scores.ts
+++ b/scripts/get-scores.ts
@@ -4,6 +4,7 @@
 // 1. Adjust the language, date, and number of scores
 // 2. npx ts-node scripts/get-scores.ts
 
+import { oldestDate } from '../src/consts';
 import { Language } from '../src/DataPopulator';
 import GitHub from '../src/GitHub';
 import StackOverflow from '../src/StackOverflow';
@@ -17,8 +18,6 @@ const languageName = 'CoffeeScript';
 const dateString = '2024-05-01';
 // Get this many scores total, starting with the above date and then getting older scores
 const numScores = 1;
-
-const oldestDate = new Date('2008-02-01');
 
 const main = async () => {
   const language = getLanguage(languageName);

--- a/scripts/get-scores.ts
+++ b/scripts/get-scores.ts
@@ -13,7 +13,7 @@ import { convertDateToDateString, subtractMonthsUTC } from '../src/utils';
 
 import 'dotenv/config';
 
-const languageName = 'ColdFusion';
+const languageName = 'CoffeeScript';
 const dateString = '2024-05-01';
 // Get this many scores total, starting with the above date and then getting older scores
 const numScores = 1;

--- a/scripts/get-scores.ts
+++ b/scripts/get-scores.ts
@@ -13,12 +13,12 @@ import { convertDateToDateString, subtractMonthsUTC } from '../src/utils';
 
 import 'dotenv/config';
 
-const oldestDate = new Date('2008-02-01');
-
 const languageName = 'ColdFusion';
-const dateString = '2023-09-01';
+const dateString = '2024-05-01';
 // Get this many scores total, starting with the above date and then getting older scores
-const numScores = 2;
+const numScores = 1;
+
+const oldestDate = new Date('2008-02-01');
 
 const main = async () => {
   const language = getLanguage(languageName);

--- a/src/DataPopulator.test.ts
+++ b/src/DataPopulator.test.ts
@@ -3,7 +3,11 @@
 import { readFile, rm, writeFile } from 'fs/promises';
 
 import DataPopulator, { Language, Score } from './DataPopulator';
-import { addMonthsUTC } from './utils';
+import {
+  addMonthsUTC,
+  getFirstDayOfMonthUTC,
+  subtractMonthsUTC,
+} from './utils';
 
 const CONDENSED_SCORES_FILE = 'scores-condensed-test.json';
 const LANGUAGES_FILE = 'languages-test.json';
@@ -17,6 +21,13 @@ const SCORES_FILE = 'scores-test.json';
 const TIME_TO_GET_ONE_SCORE = 2000;
 
 let dataPopulator: DataPopulator;
+
+beforeAll(async () => {
+  // Clean up any test files left over from previous tests; "force: true" will ignore errors
+  await rm(LANGUAGES_FILE, { force: true });
+  await rm(SCORES_FILE, { force: true });
+  await rm(CONDENSED_SCORES_FILE, { force: true });
+});
 
 describe('Tests with generated languages file', () => {
   beforeAll(() => {
@@ -51,41 +62,54 @@ describe('Tests with generated languages file', () => {
   );
 
   afterAll(async () => {
-    // Wipe scores and languages files for a clean slate
+    // Clean up to avoid an impact on other tests that use the same files
     await rm(LANGUAGES_FILE);
     await rm(SCORES_FILE);
   });
 });
 
+// These tests need the languages file to be hard-coded for more predictability since the
+// languages from GitHub will in theory change order as their popularity changes
 describe('Tests with hard-coded languages file', () => {
-  beforeAll(async () => {
-    dataPopulator = new DataPopulator(new Date('2023-01-01'));
+  beforeAll(
+    async () => {
+      // Set the oldest date based on number of scores and languages to make sure we get enough scores
+      dataPopulator = new DataPopulator(
+        subtractMonthsUTC(
+          getFirstDayOfMonthUTC(),
+          NUM_SCORES / NUM_LANGUAGES - 1
+        )
+      );
 
-    // Now hard-code the language file for more predictability for the other tests, since
-    // the languages from GitHub will in theory change order as their popularity changes
-    await writeFile(
-      LANGUAGES_FILE,
-      JSON.stringify([
-        { id: 1, name: 'C' },
-        { id: 2, name: 'C#' },
-        { id: 3, name: 'C++' },
-        { id: 4, name: 'CoffeeScript' },
-        { id: 5, name: 'Dart' },
-      ])
-    );
-    // This is still needed to populate the languages from the file in the object
-    await dataPopulator.populateLanguages(LANGUAGES_FILE, NUM_LANGUAGES);
+      await writeFile(
+        LANGUAGES_FILE,
+        JSON.stringify([
+          { id: 1, name: 'C' },
+          { id: 2, name: 'C#' },
+          { id: 3, name: 'C++' },
+          { id: 4, name: 'CoffeeScript' },
+          { id: 5, name: 'Dart' },
+        ])
+      );
+      // This is still needed to populate the languages from the file in the object
+      await dataPopulator.populateLanguages(LANGUAGES_FILE, NUM_LANGUAGES);
 
-    // Re-create the scores file
-    await dataPopulator.populateAllScores(SCORES_FILE, NUM_SCORES);
-  }, 10000);
+      // Re-create the scores file
+      await dataPopulator.populateAllScores(SCORES_FILE, NUM_SCORES);
+    },
+    // Adjust test timeout based on number of scores we're getting
+    TIME_TO_GET_ONE_SCORE * NUM_SCORES
+  );
 
   test('Test populateCondensedScores', async () => {
     await dataPopulator.populateCondensedScores(CONDENSED_SCORES_FILE);
     const scores = JSON.parse(
       await readFile(CONDENSED_SCORES_FILE, 'utf8')
     ) as Score[];
-    // The CoffeeScript scores shouldn't be included in the file
+
+    // The score for a language is typically calculated per month and then adding that
+    // value to the previous month. Because of this, the scores for CoffeeScript will be
+    // so low that they shouldn't be included in the file
     expect(scores.length).toEqual(8);
   });
 

--- a/src/DataPopulator.test.ts
+++ b/src/DataPopulator.test.ts
@@ -18,43 +18,47 @@ const TIME_TO_GET_ONE_SCORE = 2000;
 
 let dataPopulator: DataPopulator;
 
-beforeAll(async () => {
-  dataPopulator = new DataPopulator(OLDEST_DATE);
-});
+describe('Tests with generated languages file', () => {
+  beforeAll(() => {
+    dataPopulator = new DataPopulator(OLDEST_DATE);
+  });
 
-afterAll(async () => {
-  await rm(LANGUAGES_FILE);
-  await rm(SCORES_FILE);
-  await rm(CONDENSED_SCORES_FILE);
-});
+  test('Test populateLanguages', async () => {
+    await dataPopulator.populateLanguages(LANGUAGES_FILE, NUM_LANGUAGES);
+    const languages = JSON.parse(
+      await readFile(LANGUAGES_FILE, 'utf8')
+    ) as Language[];
+    expect(languages.length).toBe(NUM_LANGUAGES);
+  });
 
-test('Test populateLanguages', async () => {
-  await dataPopulator.populateLanguages(LANGUAGES_FILE, NUM_LANGUAGES);
-  const languages = JSON.parse(
-    await readFile(LANGUAGES_FILE, 'utf8')
-  ) as Language[];
-  expect(languages.length).toBe(NUM_LANGUAGES);
-});
+  test(
+    'Test populateAllScores',
+    async () => {
+      await dataPopulator.populateAllScores(SCORES_FILE, NUM_SCORES);
+      const scores = JSON.parse(await readFile(SCORES_FILE, 'utf8')) as Score[];
+      expect(scores.length).toEqual(NUM_SCORES);
+      expect(scores[0].points).toBeGreaterThan(1000);
 
-test(
-  'Test populateAllScores',
-  async () => {
-    // First make sure it works with the generated languages file
-    await dataPopulator.populateAllScores(SCORES_FILE, NUM_SCORES);
-    const scores = JSON.parse(await readFile(SCORES_FILE, 'utf8')) as Score[];
-    expect(scores.length).toEqual(NUM_SCORES);
-    expect(scores[0].points).toBeGreaterThan(1000);
+      // The latest score should be the oldest date + 1 month
+      expect(new Date(scores[scores.length - 1].date).getUTCMonth()).toEqual(
+        addMonthsUTC(OLDEST_DATE, 1).getUTCMonth()
+      );
+      // Scores should always be from the first day of the month
+      expect(new Date(scores[0].date).getUTCDate()).toEqual(1);
+    },
+    // Adjust test timeout based on number of scores we're getting
+    TIME_TO_GET_ONE_SCORE * NUM_SCORES
+  );
 
-    // The latest score should be the oldest date + 1 month
-    expect(new Date(scores[scores.length - 1].date).getUTCMonth()).toEqual(
-      addMonthsUTC(OLDEST_DATE, 1).getUTCMonth()
-    );
-    // Scores should always be from the first day of the month
-    expect(new Date(scores[0].date).getUTCDate()).toEqual(1);
-
+  afterAll(async () => {
     // Wipe scores and languages files for a clean slate
     await rm(LANGUAGES_FILE);
     await rm(SCORES_FILE);
+  });
+});
+
+describe('Tests with hard-coded languages file', () => {
+  beforeAll(async () => {
     dataPopulator = new DataPopulator(new Date('2023-01-01'));
 
     // Now hard-code the language file for more predictability for the other tests, since
@@ -74,56 +78,60 @@ test(
 
     // Re-create the scores file
     await dataPopulator.populateAllScores(SCORES_FILE, NUM_SCORES);
-  },
-  // Adjust test timeout based on number of scores we're getting
-  TIME_TO_GET_ONE_SCORE * NUM_SCORES
-);
+  }, 10000);
 
-test('Test populateCondensedScores', async () => {
-  await dataPopulator.populateCondensedScores(CONDENSED_SCORES_FILE);
-  const scores = JSON.parse(
-    await readFile(CONDENSED_SCORES_FILE, 'utf8')
-  ) as Score[];
-  // The CoffeeScript scores shouldn't be included in the file
-  expect(scores.length).toEqual(8);
-});
+  test('Test populateCondensedScores', async () => {
+    await dataPopulator.populateCondensedScores(CONDENSED_SCORES_FILE);
+    const scores = JSON.parse(
+      await readFile(CONDENSED_SCORES_FILE, 'utf8')
+    ) as Score[];
+    // The CoffeeScript scores shouldn't be included in the file
+    expect(scores.length).toEqual(8);
+  });
 
-test('Test validateLanguages', async () => {
-  try {
-    await dataPopulator.validateLanguages(LANGUAGES_FILE);
-  } catch (error) {
-    console.log(error);
-  }
-});
+  test('Test validateLanguages', async () => {
+    try {
+      await dataPopulator.validateLanguages(LANGUAGES_FILE);
+    } catch (error) {
+      console.log(error);
+    }
+  });
 
-test(
-  'Test significant decrease in points',
-  async () => {
-    // Suppress logs from this test (https://stackoverflow.com/a/58717352/399105)
-    jest.spyOn(console, 'debug').mockImplementation(() => {});
-    jest.spyOn(console, 'info').mockImplementation(() => {});
+  test(
+    'Test significant decrease in points',
+    async () => {
+      // Suppress logs from this test (https://stackoverflow.com/a/58717352/399105)
+      jest.spyOn(console, 'debug').mockImplementation(() => {});
+      jest.spyOn(console, 'info').mockImplementation(() => {});
 
-    // Wipe the scores file for a clean slate and then re-create it
+      // Wipe the scores file for a clean slate and then re-create it
+      await rm(SCORES_FILE);
+
+      // Populate two months of data so we can manipulate the second to most recent one
+      await dataPopulator.populateAllScores(SCORES_FILE, NUM_LANGUAGES * 2);
+
+      // Set a high score for CoffeeScript
+      const scores = JSON.parse(await readFile(SCORES_FILE, 'utf8')) as Score[];
+      scores[3].points = 2000;
+
+      // Delete the most recent scores so they'll be repopulated
+      scores.splice(NUM_LANGUAGES);
+
+      // Overwrite the scores file
+      await writeFile(SCORES_FILE, JSON.stringify(scores));
+
+      // An error should be thrown because of the significant decrease in points
+      await expect(
+        dataPopulator.populateAllScores(SCORES_FILE, NUM_SCORES)
+      ).rejects.toThrow('decreased');
+    },
+    // Adjust test timeout based on number of scores we're getting
+    TIME_TO_GET_ONE_SCORE * NUM_SCORES * 2
+  );
+
+  afterAll(async () => {
+    await rm(LANGUAGES_FILE);
     await rm(SCORES_FILE);
-
-    // Populate two months of data so we can manipulate the second to most recent one
-    await dataPopulator.populateAllScores(SCORES_FILE, NUM_LANGUAGES * 2);
-
-    // Set a high score for CoffeeScript
-    const scores = JSON.parse(await readFile(SCORES_FILE, 'utf8')) as Score[];
-    scores[3].points = 2000;
-
-    // Delete the most recent scores so they'll be repopulated
-    scores.splice(NUM_LANGUAGES);
-
-    // Overwrite the scores file
-    await writeFile(SCORES_FILE, JSON.stringify(scores));
-
-    // An error should be thrown because of the significant decrease in points
-    await expect(
-      dataPopulator.populateAllScores(SCORES_FILE, NUM_SCORES)
-    ).rejects.toThrow('decreased');
-  },
-  // Adjust test timeout based on number of scores we're getting
-  TIME_TO_GET_ONE_SCORE * NUM_SCORES * 2
-);
+    await rm(CONDENSED_SCORES_FILE);
+  });
+});

--- a/src/DataPopulator.ts
+++ b/src/DataPopulator.ts
@@ -3,6 +3,7 @@
 import { readFile, writeFile } from 'fs/promises';
 import fetch from 'node-fetch';
 
+import { oldestDate } from './consts';
 import GitHub from './GitHub';
 import _languagesMetadata from '../data/languages-metadata.json';
 import settings from './settings.json';
@@ -45,16 +46,14 @@ export default class DataPopulator {
   private github: GitHub;
   private languages: Language[];
   private languagesFromGithub?: string[];
+  private oldestDate: Date;
   private scores: Score[];
   private stackoverflow: StackOverflow;
 
-  // The oldest date with data is 2007-11-01 but no languages have a score > 1 before 2008-02-01
-  private oldestDate = new Date('2008-02-01');
-
   /**
-   * @param oldestDate - Oldest date to use for populating languages (used for testing)
+   * @param oldestDateOverride - Oldest date to use for populating languages (used for testing)
    */
-  constructor(oldestDate?: Date) {
+  constructor(oldestDateOverride?: Date) {
     if (!process.env.GITHUB_API_KEY) {
       throw new Error('GITHUB_API_KEY must be set');
     }
@@ -62,7 +61,7 @@ export default class DataPopulator {
     this.firstDayOfMonth = getFirstDayOfMonthUTC();
     this.github = new GitHub(process.env.GITHUB_API_KEY);
     this.languages = [];
-    this.oldestDate = oldestDate ?? this.oldestDate;
+    this.oldestDate = oldestDateOverride ?? oldestDate;
     this.scores = [];
     this.stackoverflow = new StackOverflow(process.env.STACKOVERFLOW_API_KEY);
   }

--- a/src/DataPopulator.ts
+++ b/src/DataPopulator.ts
@@ -10,6 +10,7 @@ import StackOverflow from './StackOverflow';
 import {
   addMonthsUTC,
   convertDateToDateString,
+  getFirstDayOfMonthUTC,
   subtractMonthsUTC,
 } from './utils';
 
@@ -58,7 +59,7 @@ export default class DataPopulator {
       throw new Error('GITHUB_API_KEY must be set');
     }
 
-    this.firstDayOfMonth = DataPopulator.getFirstDayOfMonthUTC();
+    this.firstDayOfMonth = getFirstDayOfMonthUTC();
     this.github = new GitHub(process.env.GITHUB_API_KEY);
     this.languages = [];
     this.oldestDate = oldestDate ?? this.oldestDate;
@@ -182,12 +183,6 @@ export default class DataPopulator {
     this.sortScores();
 
     await writeFile(scoresFile, JSON.stringify(this.scores, null, 2));
-  }
-
-  private static getFirstDayOfMonthUTC(): Date {
-    return new Date(
-      Date.UTC(new Date().getUTCFullYear(), new Date().getUTCMonth())
-    );
   }
 
   private async populateScoresForDate(date: Date, numScores?: number) {
@@ -340,7 +335,7 @@ export default class DataPopulator {
   public async populateCondensedScores(condensedScoresFile: string) {
     // Get settings from the frontend
     const response = await fetch(
-      'https://raw.githubusercontent.com/bmaupin/langtrends/master/src/settings.json'
+      'https://raw.githubusercontent.com/bmaupin/langtrends/main/src/settings.json'
     );
     const uiSettings = await response.json();
 

--- a/src/StackOverflow.ts
+++ b/src/StackOverflow.ts
@@ -68,7 +68,7 @@ export default class StackOverflow {
     return url;
   }
 
-  private async callApi(url: string) {
+  private async callApi(url: string): Promise<StackOverflowData> {
     const options = new URL(url);
     const bodyJson = await this.httpsRequest(options);
     return JSON.parse(bodyJson);
@@ -106,6 +106,9 @@ export default class StackOverflow {
               zlib.gunzip(Buffer.concat(body), (error, uncompressedData) => {
                 resolve(uncompressedData.toString());
               });
+              break;
+            case undefined:
+              resolve(Buffer.concat(body).toString());
               break;
             default:
               // If we get here it's likely due to another issue, normally a 503 error due to too many requests

--- a/src/consts.ts
+++ b/src/consts.ts
@@ -1,0 +1,2 @@
+// The oldest date with data is 2007-11-01 but no languages have a score > 1 before 2008-02-01
+export const oldestDate = new Date('2008-02-01');

--- a/src/utils.ts
+++ b/src/utils.ts
@@ -15,6 +15,14 @@ export const convertDateToDateString = (date: Date): string => {
   return date.toISOString().slice(0, 10);
 };
 
+// Get a date representing the first day of the current month based on UTC; that is to say
+// if the local day is 2024-03-31 and UTC is 2024-04-01, it should return 2024-04-01
+export const getFirstDayOfMonthUTC = (): Date => {
+  return new Date(
+    Date.UTC(new Date().getUTCFullYear(), new Date().getUTCMonth())
+  );
+};
+
 // Source: https://github.com/bmaupin/langtrends/blob/master/src/helpers/ApiHelper.js
 export const subtractMonthsUTC = (
   date: Date,

